go/extractor/extractor.go

```diff
@@ -936,7 +941,16 @@ func emitScopeNodeInfo(tw *trap.Writer, nd ast.Node, lbl trap.Label) {
 
 // extractExpr extracts AST information for the given expression and all its subexpressions
 func extractExpr(tw *trap.Writer, expr ast.Expr, parent trap.Label, idx int, skipExtractingValue bool) {
-	if expr == nil {
+	if expr == nil || expr == (*ast.Ident)(nil) || expr == (*ast.BasicLit)(nil) ||
```

Comment:
TIL: this makes sense. It does, though! Weak preference for keeping the old code, since as/when there are new AST nodes it is now easier to accidentally not add the nil check.

Reply:
as a Go neophyte, after doing a quick google search, would something like `expr == nil || reflect.ValueOf(i).IsNil()` work and avoid the problem?

Reply:
It's a definite footgun in the language, which I think they probably should have a better solution for. Btw there is an alternative which I saw suggested when I searched for this online, which is to use reflection to check if the interface type has a nil value. I thought this was too ugly, but it would address the concern that we need to remember to add any new types to the check.

You want to leave in the nil checks in the branches of the type switch? I don't really agree, but I'll do it as it isn't worth debating.

Reply:
My comment above was a response to @smowton 's comment. I'm not sure about the exact formulation of using reflection. Since we know it will be an interface and not a map or channel or anything like that, it shouldn't be too hard.

Reply:
If the reflection isn't too costly that's fine. After that I'd say a check in each branch is slightly better the a mass of checks up front, because the up-front ones can drift out of sync with the arms of the switch too easily.

Reply:
> It's a definite footgun in the language

So true! TIL, I'll definitely keep that in mind for next time I might dabble with Go.

Reply:
Another idea occurred to me: we just move the two lines that are above the type switch into every case clause. Then the nil check works as expected. There is duplication, but it isn't too bad, and anyone adding another case in future will be able to copy from other case clauses without any risk of error. I think if reflection works and is performant then that is probably my first choice, followed by this duplication approach.